### PR TITLE
CHANGE (GraphEngine): @W-12423302@: Excludes evaluating paths that have stack depth over the allowed Apex governor limit.

### PR DESCRIPTION
**Pull request by rmohan20**

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java [hunk 1]
```diff
@@ -92,6 +96,8 @@
     /** Dynamically generated id used to establish object equality */
     private final Long id;
 
+    private int stackDepth;
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java:99]: Maintaining a stack depth count within ApexPathExpander.

File: sfge/src/main/java/com/salesforce/config/EnvUtil.java [hunk 2]
```diff
@@ -76,6 +79,14 @@ static int getProgressIncrements() {
         return getIntOrDefault(ENV_PROGRESS_INCREMENTS, DEFAULT_PROGRESS_INCREMENTS);
     }
 
+    /**
+     * Returns stack depth limit that Apex's governor limits enforce. Cannot be overridden by a
+     * user-specified value.
+     */
+    static int getStackDepthLimit() {
+        return DEFAULT_STACK_DEPTH;
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/config/EnvUtil.java:87]: Not overridable through config.

File: sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java [hunk 3]
```diff
@@ -641,12 +653,16 @@ public void visit(AbstractPathBasedRule rule) {
                 }
             }
 
+            // Increment stack depth before visiting a method. Decrement stack depth when returning.
+            incrementStackDepth(vertex);
             visit(methodPath);
+            decrementStackDepth(vertex);
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpander.java:659]: Here's where all action happens.

File: sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderUtil.java [hunk 4]
```diff
@@ -274,6 +274,11 @@ private void expand(
                     if (LOGGER.isWarnEnabled()) {
                         LOGGER.warn("expand-NullAccess. ex=" + ex);
                     }
+                } catch (StackDepthLimitExceededException ex) {
+                    apexPathCollapser.removeExistingExpander(apexPathExpander);
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathExpanderUtil.java:278]: Exclude paths that hit the stack depth limit.

File: sfge/src/test/java/com/salesforce/graph/ops/expander/StackDepthLimitTest.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    @BeforeEach
+    public void setup() {
+        this.g = TestUtil.getGraph();
+        SfgeConfigTestProvider.set(
+                new TestSfgeConfig() {
+                    @Override
+                    public int getStackDepthLimit() {
+                        return STACK_DEPTH_LIMIT_OVERRIDE;
+                    }
```
Comment (rmohan20) [inline sfge/src/test/java/com/salesforce/graph/ops/expander/StackDepthLimitTest.java:33]: Using an overridden stack depth value for relevant tests.

File: sfge/src/test/java/com/salesforce/graph/ops/expander/StackDepthLimitTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        "WithinLimit, " + (STACK_DEPTH_LIMIT_OVERRIDE - 2) + ", true"
+    })
+    @ParameterizedTest(name = "{displayName}: {0}")
+    public void testStackDepth(
+            String testName, int stackDepthToCreate, boolean expectToReachSystemDebug) {
+        // Since we already create two methods other than the dynamic ones, subtract them from
+        // required count
+        final int depthCountToAdd = stackDepthToCreate - 2;
+        final String sourceCode =
+                "public class MyClass {\n"
```
Comment (jfeingold35) [inline sfge/src/test/java/com/salesforce/graph/ops/expander/StackDepthLimitTest.java:63]: You can wrap your code in `spotless:off` and `spotless:on` to make it immune to reformatting. I find it makes the source code in tests way easier to read.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**